### DISCUSSION TITLE
Set kb_id to "" (empty string; id 0) instead of NIL for no prediction in entity_linker

### DISCUSSION THREAD

**Discussion by kinghuang** [explosion/spaCy#9580] [New Features & Project Ideas]
When there is no prediction or an entity is discarded, the [EntityLinker](https://spacy.io/api/entitylinker) sets the kb_id to `NIL`, instead of an empty string. This means that there are two values for no `kb_id`: an empty string (id 0) or `NIL`.

https://github.com/explosion/spaCy/blob/5279c7c4baf5b9f26a364f07551acecf26a2d42f/spacy/pipeline/entity_linker.py#L105

https://github.com/explosion/spaCy/blob/5279c7c4baf5b9f26a364f07551acecf26a2d42f/spacy/pipeline/entity_linker.py#L331-L332

I don't think the entity_linker should be using a special string, `NIL`, to represent no `kb_id`. This complicates code, which can't depend on `Span.kb_id` or `Span.kb_id_` evaluating to a falsy value when there is no `kb_id`. And, it differs from entity labels, where it is consistently an empty string for no label.

**Comment by svlandeg** [comment]
Hi,

I think our original reasoning was that there's a difference between "not set" and "no link possible". For instance, I myself am not represented in Wikidata, so any text including a reference to my name, should get a gold annotation `NIL` to specifically state there is no suitable link.

**Reply by kinghuang** [reply]
Ah, I see. However, I'd argue this is no different than the EntityRecognizer not having a label to apply to a range of tokens. There's no distinction between "not set" and "no entity" on `Span.label_`. However, there is a different on the corresponding `Token.ent_iob_` attribute (empty vs `O`).

Can the `NIL` be kept on the `Token.ent_kb_id_` attribute, but an empty value on the `Span.kb_id_` attribute, like how entity labels are handled?

**Reply by svlandeg** [reply]
I think right now we don't want to change the behaviour, but I'll put this in the backlog to (re)consider in the future!

**Reply by kinghuang** [reply]
Ok. FYI, I have run into a complication with the `NIL` kb_id when using multiple `entity_linker` components with `overwrite=False`. The entity linker will only set kb_id when it is empty/`0`, not when it's `NIL`.

https://github.com/explosion/spaCy/blob/86fa37e8baf631348ec712a174c19c3ca7fb88cd/spacy/pipeline/entity_linker.py#L416

**Reply by svlandeg** [reply]
Thanks, that's useful to add to the consideration! (I'm sure you've figured that out, but a quick workaround is a custom component inbetween the two EL's that sets NIL's back to empty...)
